core/src/main/java/org/apache/iceberg/io/CDCWriter.java

```diff
@@ -0,0 +1,135 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.io;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.function.Function;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.deletes.PositionDelete;
+import org.apache.iceberg.util.StructLikeMap;
+import org.apache.iceberg.util.StructProjection;
+
+public class CDCWriter<T> extends BaseDeltaWriter<T> {
+
+  private final FanoutDataWriter<T> dataWriter;
+  private final PartitionAwareFileWriter<T, DeleteWriteResult> equalityDeleteWriter;
+  private final PartitionAwareFileWriter<PositionDelete<T>, DeleteWriteResult> positionDeleteWriter;
+  private final StructProjection keyProjection;
+  private final Map<StructLike, PartitionAwarePathOffset> insertedRows;
+  private final PositionDelete<T> positionDelete;
+  private final Function<T, StructLike> toStructLike;
+
+  public CDCWriter(FanoutDataWriter<T> dataWriter,
+                   PartitionAwareFileWriter<T, DeleteWriteResult> equalityDeleteWriter,
+                   PartitionAwareFileWriter<PositionDelete<T>, DeleteWriteResult> positionDeleteWriter,
+                   Schema schema, Schema deleteSchema, Function<T, StructLike> toStructLike) {
+    this.dataWriter = dataWriter;
+    this.equalityDeleteWriter = equalityDeleteWriter;
+    this.positionDeleteWriter = positionDeleteWriter;
+    this.positionDelete = new PositionDelete<>();
+    this.keyProjection = StructProjection.create(schema, deleteSchema);
+    this.insertedRows = StructLikeMap.create(deleteSchema.asStruct());
+    this.toStructLike = toStructLike;
+  }
+
+  @Override
+  public void insert(T row, PartitionSpec spec, StructLike partition) throws IOException {
+    CharSequence currentPath = dataWriter.currentPath(spec, partition);
+    long currentPosition = dataWriter.currentPosition(spec, partition);
+    PartitionAwarePathOffset offset = new PartitionAwarePathOffset(spec, partition, currentPath, currentPosition);
+
+    StructLike copiedKey = StructCopy.copy(keyProjection.wrap(toStructLike.apply(row)));
+
+    PartitionAwarePathOffset previous = insertedRows.put(copiedKey, offset);
+    if (previous != null) {
+      // TODO: attach the previous row if has a position delete row schema
+      positionDelete.set(previous.path(), previous.rowOffset(), null);
+      positionDeleteWriter.write(positionDelete, spec, partition);
+    }
+
+    dataWriter.write(row, spec, partition);
+  }
+
+  @Override
+  public void delete(T row, PartitionSpec spec, StructLike partition) throws IOException {
```

Comment:
I think we need to clearly document that this assumes that the row to delete has the same schema as the rows that will be inserted. We could also have a `directDelete` method that passes just the equality delete columns (key). That's worth considering if you want to split the interface for equality and position delete use cases.